### PR TITLE
Added diff_state st+1-st

### PR DESCRIPTION
**Pull request by sdk-bonsai**
# Pull Request Template

## Description

This is a small PR.  I added a config variable flag `diff_state` to all configs. This flag is used to reassign output  states `st+1` as a difference compared to last state `st+1-st`. Changed `ddm_test_validate.py` and `ddm_predictor.py` and `base.py` accordingly.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X ] This change requires a documentation update

## How Has This Been Tested?

Tested with the `ddm_test_validate.py`.

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Reviewer Tasks

List here all the tasks the reviewer needs to review:

- [ ] Test `examples/`

**Test Configuration**:
* OS:
* Python Version:
* Bonsai SDK Version:

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: conf/model/torch.yaml [hunk 1]
```diff
@@ -1,25 +1,26 @@
...
   name: pytorch
   build_params:
-    num_units: 50
-    dropout: 0.5
-    num_layers: 10
-    device: cpu
-    batch_size: 128
-    num_epochs: 10
-    scale_data: True
+    - network_class: MVRegressor
```
Comment (akzaidi) [inline conf/model/torch.yaml:4]: please don't add the - here. they'll break the `build_model` functions.

Comment (sdk-bonsai) [reply conf/model/torch.yaml]: That actually worked for training the torch model. But, I switched back to the previous torch config.

File: ddm_predictor.py [hunk 2]
```diff
@@ -31,14 +31,23 @@
 
 
 class Simulator(BaseModel):
-    def __init__(self, model, states=List[str], actions=List[str], configs=List[str]):
+    def __init__(
+        self,
+        model,
+        states=List[str],
```
[the remark this answers is not in the document]
Comment (sdk-bonsai) [reply ddm_predictor.py]: Changed to colons, same change in ddm_test_validate also changed `model.build_model(**cfg["model"]["build_params"])`

File: ddm_predictor.py [hunk 3]
```diff
@@ -31,14 +31,23 @@
...
 
 class Simulator(BaseModel):
-    def __init__(self, model, states=List[str], actions=List[str], configs=List[str]):
+    def __init__(
+        self,
+        model,
+        states: List[str],
+        actions: List[str],
+        configs: List[str],
+        log_file: str = None,
```
Comment (akzaidi) [inline ddm_predictor.py:40]: The parameter `log_file` is never accessed.

**Review by akzaidi — COMMENTED** [review]
we need to add the argument `del_state` in `ddm_trainer.py` if we want to use `delta_states` during training right?

```python
    X, y = model.load_csv(
        input_cols=input_cols,
        output_cols=output_cols,
        augm_cols=augmented_cols,
        dataset_path=dataset_path,
        iteration_order=iteration_order,
        episode_col=episode_col,
        iteration_col=iteration_col,
        max_rows=max_rows,
        ... add del_state=cfg["data"]["diff_state"] ?
    )
```



​also do you mind if we standardize the name to diff_state or delta_state? it's diff_state in the config but the variable is named del_state , which is a bit confusing (`delta_state` might be better overall?)

**Comment by akzaidi** [comment]
LGTM. Small consistency changes and update the docstring for the new param, and noted the TODO for logging. Tested:

```bash
python ddm_trainer.py data.diff_state=True
```

Logs:
```bash
[2021-04-01 12:05:31,987][datamodeler][INFO] - Configuration: 
[2021-04-01 12:05:31,990][datamodeler][INFO] - 
data:
  path: csv_data/cartpole-log.csv
  full_or_relative: relative
  inputs: state
  outputs: state
  augmented_cols:
  - config_length
  - config_masspole
  - action_command
  iteration_order: -1
  episode_col: episode
  iteration_col: iteration
  max_rows: 1000
  scale_data: true
  diff_state: true
model:
  name: xgboost
  build_params:
    model_type: xgboost
    halt_model: false
    num_trees: 50
    objective: reg:squarederror
    step_size: 0.3
    device: cpu
    batch_size: 128
    gamma: 0
    scale_data: true
    max_bin: 256
  saver:
    filename: models/boost/moab/xgboost_model
  sweep:
    run: false
    search_algorithm: bayesian
    num_trials: 3
    scoring_func: r2
simulator:
  states:
  - state_x_position
  - state_x_velocity
  - state_angle_position
  - state_angle_velocity
  actions:
  - action_command
  configs:
  - config_length
  - config_masspole
  policy: bonsai
  logging: enable

[2021-04-01 12:05:31,997][root][INFO] - Using ['state_x_position', 'state_x_velocity', 'state_angle_position', 'state_angle_velocity', 'config_length', 'config_masspole', 'action_command'] as the features for modeling DDM
[2021-04-01 12:05:31,997][root][INFO] - Using ['state_x_position', 'state_x_velocity', 'state_angle_position', 'state_angle_velocity'] as the labels for modeling DDM
[2021-04-01 12:05:31,997][data_loaders][INFO] - Iteration order set to -1 so using inputs from previous -1 row
[2021-04-01 12:05:32,002][data_loaders][INFO] - Previous states are being added to same row with prefix prev
[2021-04-01 12:05:32,002][data_loaders][INFO] - Feature columns are: ['prev_state_x_position', 'prev_state_x_velocity', 'prev_state_angle_position', 'prev_state_angle_velocity', 'prev_config_length', 'prev_config_masspole', 'prev_action_command']
[2021-04-01 12:05:32,002][data_loaders][INFO] - Label columns are: ['state_x_position', 'state_x_velocity', 'state_angle_position', 'state_angle_velocity']
[2021-04-01 12:05:32,018][root][INFO] - delta states enabled, calculating differential between current and previous predicted output
[2021-04-01 12:05:32,018][datamodeler][INFO] - Building model...
[2021-04-01 12:05:32,019][datamodeler][INFO] - Fitting model...
[2021-04-01 12:05:32,569][datamodeler][INFO] - Saving model to /Users/alizaidi-msft/Documents/bonsai/datadrivenmodel/models/boost/moab/xgboost_model.pkl
[2021-04-01 12:05:32,569][gboost_models][INFO] - Scale transformations used, saving to /Users/alizaidi-msft/Documents/bonsai/datadrivenmodel/models/boost/moab/xgboost_model.pkl

```

**Pull request merged**